mono/mini/intrinsics.c

```diff
@@ -1139,6 +1139,49 @@ mini_emit_inst_for_method (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSign
 		 * all inputs:
 		 * http://everything2.com/?node_id=1051618
 		 */
+		if (cmethod->klass == mono_defaults.math_class){
+			/* Constant folding for Math.{Sin,Cos,Tan,Sqrt} */
+			if (fsig->param_count == 1 && args [0]->opcode == OP_R8CONST){
+				int opcode = 0;
+				if (strcmp (cmethod->name, "Sin") == 0) {
+					opcode = OP_SIN;
+				} else if (strcmp (cmethod->name, "Cos") == 0) {
+					opcode = OP_COS;
+				} else if (strcmp (cmethod->name, "Sqrt") == 0) {
+					opcode = OP_SQRT;
+				} else if (strcmp (cmethod->name, "Abs") == 0 && fsig->params [0]->type == MONO_TYPE_R8) {
+					opcode = OP_ABS;
+				}
+
+				if (opcode){
+					double *dest = (double *) mono_domain_alloc (cfg->domain, sizeof (double));
+					double source = *(double *)args [0]->inst_p0;
+					double result;
+					MONO_INST_NEW (cfg, ins, OP_R8CONST);
+					ins->type = STACK_R8;
+					ins->dreg = mono_alloc_dreg (cfg, ins->type);
+					ins->inst_p0 = dest;
+					
+					switch (opcode){
+					case OP_SQRT:
+						result = sqrt (source);
+						break;
+					case OP_COS:
+						result = cos (source);
+						break;
+					case OP_SIN:
+						result = sin (source);
+						break;
+					case OP_ABS:
+						result = fabs (source);
+						break;
+					}
+					*dest = result;
```

Comment:
Hopefully these will return the same value on the host and target machines when cross compiling.

Reply:
They are constants, so they should be the same.

I wanted to get your opinion generally on the code so far that adds the capability.

Reply:
There could be precision problems between the implementations on the different platforms. Also, some functions are not defined for some values, so some error handling is needed here.
You might need something like:
https://en.wikipedia.org/wiki/GNU_MPFR

Reply:
You mean for the constant folding of floating point values, or the math operations that are being inlined?    I will check with RyuJIT is doing.